### ISSUE TITLE
[Language service] nullable types cannot be ==-compared to null

### ISSUE THREAD

**Issue by lemoinem** [angular/angular#18754]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
I have an interface:
```ts
export interface Foo
{
   name?: string;
}
```

In my template, I'm doing:
```html
<span *ngIf="myFoo.name != null">{{ myFoo.name }}</span>
```

I'm getting an error "Expected the operants to be of similar type or any".
Same thing if I'm trying to compare to undefined or using ===.

## Expected behavior
No error reported.

## What is the motivation / use case for changing the behavior?
Current behavior is wrong and produce numerous false positives.

## Environment

<pre><code>
Angular version: 4.3.1 / Angular Language Service 0.1.4

Browser: None - language service (VSCode)
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: v6.10.0
- NPM: 5.3.0
- Platform:  Windows 10

Others:
VSCode 1.15.0 x64 using Node 7.4.0
</code></pre>

**Comment by Chris3773** [comment]
The question mark on the property tells TypeScript that the value can be undefined not nullable. You would need to change the property to `name?: string | null;` to allow undefined and null values.

**Comment by lemoinem** [comment]
@Chris3773 I agree, but as I said, changing my test for `myFoo.name != undefined` doesn't remove the error message. So this is definitely a bug.

**Comment by rudym87** [comment]
I'm getting this error too, in my case I'm checking if an optional string property is present in order to assign a CSS class, but i do want to apply the class even though the string is empty.
````html
<div [class.time-slot__message]="time?.msg !== undefined"></div>
````
The error shows only in VS Code, I don't get any error/warning during `ng lint` or build.

VS Code 1.18.1
Angular Language Service Extension 0.1.7

**Comment by youssefabidi** [comment]
I have the same error message (_Expected the operants to be of similar type or any_) in VS Code (with the Angular Language Service) in a very similar situation (the nullable property is just a number instead of a string and is defined in a class instead of an interface):
```html
<div *ngIf="item.size !== null"></div>
```
Where `item` is an instance of the class:
```typescript
export class Item {
    size: number | null;
}
```
The error message is for the condition of the `*ngIf`. It is only present in VS Code and does not happen during build.

**Comment by aj-r** [comment]
I think this is a duplicate of #22125. Looks like the problem happens for all union types, not just nullable types.

**Comment by JoostK** [comment]
This was fixed in #36529 and also won't be an issue anymore with the Ivy language service.
